domain.c: use an atomic counter for domain unique ids

gasche opened the pull request:
The previous implementation would increment each domain's identifier
by Max_domain. We want to remove the dependence on a fixed Max_domains
setting, to be able to reconfigure it during program execution (see 
https://github.com/ocaml/ocaml/issues/10971#issuecomment-1025507893).
    
Note: the previous low-level domain-synchronization code would
use `unique_id % Max_domains` to efficiently compute the domain index:
  9e9eb7972489c557d3c214dc0398a7a3e81ce2c3
but this logic was entirely removed by
  d5694dade56e912fa6d4574c6387c2f12aafa331

gasche committed:
domain.c: use an atomic counter for domain unique ids

The previous implementation would increment each domain's identifier
by Max_domain. We want to remove the dependence on a fixed Max_domains
setting, to be able to reconfigure it during program execution.

Note: the previous low-level domain-synchronization code would
use (unique_id % Max_domains) to efficiently compute the domain index:
  9e9eb7972489c557d3c214dc0398a7a3e81ce2c3
but this logic was entirely removed by
  d5694dade56e912fa6d4574c6387c2f12aafa331

gasche committed:
domain: simplify the timing of fresh unique_id creation

The previous implementation would pre-compute domain unique ids before
the domains were created -- when initializing Max_domains at startup,
and when terminating the previous domain using the same slot. This is
not necessary anymore.

gasche committed:
[minor] domain.c: document domain_startup_params in/out status

gasche commented:
Note: there is something fishy with the type of `unique_id` fields: it is `uintnat` in domain.c, but `int` in caml/domain.tbl eventlog.c and even `uint32_t` in the FWRITE_HEADER macro of eventlog.c. Should we use `uintnat` everywhere, and `uint64_t` for eventlog serialization?

kayceesrk commented:
We hold `all_domains_lock` in `create_domain`. https://github.com/ocaml/ocaml/blob/6a7a7b36fa2b6f4c4f3d210c8ecaff6faa094adf/runtime/domain.c#L397-L399 

So we may make `next_domain_unique_id` a non atomic variable protected by the `all_domains_lock`.

> Should we use uintnat everywhere, and uint64_t for eventlog serialization?

Yes. That's the right thing to do.

gasche committed:
domain.c: next_domain_unique_id is already protected by all_domains_lock

Suggested-by: KC Sivaramakrishnan <kc@kcsrk.info>

gasche commented:
Thanks! I didn't notice this in my first iteration, as the unique identifiers were generated in other places; but then the code moves the generation to `create_domain` only and your suggestion is a nice simplification. I pushed it as a new commit.

I did not change the types in domain_state.tbl and eventlog.c; I prefer to do this separately and may look at the type of other domain-state fields at the same time.

kayceesrk reviewed (approved):
LGTM! Thank you.

ctk21 commented on runtime/domain.c:
```diff
@@ -930,7 +940,7 @@ CAMLprim value caml_domain_spawn(value callback, value mutex)
+  return Val_long(domain_self->interruptor.unique_id);
 }
 
 CAMLprim value caml_ml_domain_unique_token (value unit)
```
As I was looking at this, I noticed that `caml_ml_domain_unique_token` is a relic that can be removed:
 https://github.com/ocaml/ocaml/commit/f0e76441549719f126d4d3c7e0aafb6cd2686764

gasche replied:
Note: I tried to get rid of it naively, but the build breaks because a primitive was removed. I think my bootstrap-fu is not strong enough yet to do this easily (there is some documentation for how to do this in BOOTSTRAP.adoc, but I was hoping that it wasn't necessary to bootstrap anymore thanks to the `-use-prims` flag, which I couldn't find a noob-friendly explanation for.) From the fact that it simply returns `Val_unit`, I suspect that the person who worked on it last hit the same barrier and decided to just remove the implementation and keep the primitive. Some other time!

ctk21 reviewed (commented):
I think the motivation is sound: avoid the fixed `Max_domains`. I think the change is fine in terms of concurrency because of `all_domains_lock`. 

There is a (theoretical) nit:
 - in the old code, when the `unique_id` wrapped, the domains will have unique identifiers because a domain slot is incremented by `Max_domains` and so separation of the id cycles between slots (assuming `Max_domains` is an appropriate divisor of the maximum id). In the proposed code the id might no longer be unique. For example, on wrap there will be two domains with `unique_id == 0` and then `Domain.is_main_domain ()` will also be true in a spawned domain. 

We could avert our eyes to this; if we can spawn 1B domains a second, then it would take over 500Y to achieve the wrap with a 64-bit id space. A 32-bit id space might be problematic. Currently even bytecode doesn't run on 32-bit architectures for other reasons (e.g. default virtual memory reservation), so I'm not sure it has to be a blocker. Maybe an appropriate comment at `next_domain_unique_id++`?
 
I also checked if tests are dependent on the implementation of domain id. `tests/parallel/domain_id.ml` relies on monotonic increasing ids and so is robust to the change; however it has a comment about `Max_domains` that should be updated to match. (There are also usages of `Domain.self` in tests with ignored return!)

xavierleroy commented:
Re: overflow in domain unique IDs, we have (and have had for a long time) the same problem with thread identifiers, which are produced by incrementing a `uintnat`.  So, I guess we're used to close our eyes on 32-bit platforms...

Here, since domain 0 is special (?), we could check for overflow and skip the 0 value.

abbysmal commented:
> eventlog.c and even uint32_t in the FWRITE_HEADER macro of eventlog.c. Should we use uintnat everywhere, and uint64_t for eventlog serialization?

Good catch.
Eventlog needs some attention in a few areas, I can take care of this, but there's also #10964 which aims to be the long-term solution to the eventlog infrastructure in Multicore.
I can take care of polishing the eventlog part if you want me to (because outside of `eventlog.c`, we need to be careful with `tools/eventlog_metadata` as well.).

gasche commented:
Thanks for the overflow consideration; I agree that it's a potential concern on 32 bits.

I wanted to propose to just store the unique_id as an `uint64_t` unconditionally, as this is simple and gives better guarantees. However, this value is also exposed on the OCaml side as the type `Domain.id = private int`. If we used `uint64_t` on the C side, we would need to cast on the OCaml side or use `Int64.t`.

If we want to keep only an `int`-sized value on the OCaml side, then we probably want do the increment-if-0-again logic that @xavierleroy suggests at domain-creation time, rather than at `Domain.get_id` time.

But is it reasonable to keep domain identifiers as `int` in OCaml? I suspect that if we do want to claim that they are unique, then we should in fact ensure that they are unique by providing `Int64.t` values. This is going to be fast on 64bit machines, and there is a valid reason to pay the cost on 32bit machines.

Note: these unique identifiers are not meant to be used for array-indexing for example, as in general they do not remain in a small [0..N] range (certainly not [0..numcores]). The "index" value is `domain_state->id`, which is currently *not* exposed on the OCaml side as far as I can see. I don't see how the "unique id" could be used in performance sensitive scenarios then -- or at least they will be used with sparse map structures relying on hashing and comparison, so the performance overhead of Int64.t compared to int would probably be drowned in the noise.

So here is my proposal:
- migrate unique identifiers to use `uint64_t` on the C sid, and `Int64.t` on the OCaml side
- move `is_main_domain` into the runtime, to avoid spreading assumptions on unique identifiers between the OCaml code and the C code

gasche commented:
@Engil yes, I think it would be helpful if you could do eventlog.c changes. My gut feeling is that, on the C side, it is safer to consistently serialize them as `uint64_t`. This is certainly more correct on 64bit systems (our main target), where currently we could lose information on overflow; even on the 32bit side, this would leave us some wiggle room to change the definition (as we are discussing now) without having to change the format again.

alainfrisch commented:
> Re: overflow in domain unique IDs, we have (and have had for a long time) the same problem with thread identifiers, which are produced by incrementing a uintnat. So, I guess we're used to close our eyes on 32-bit platforms...

Not really related to the discussion here, but there is also the problem of the uid used for objects and extension constructors.  The generator can wrap on 32-bit, and the uid is what defines the notion of equality (for the generic comparison) for those values, which can lead to surprising semantics (including with js_of_ocaml).

xavierleroy commented:
Yes, 32-bit platforms have a lot of limitations.  (I told you so.)  We're going to support them a bit, or not at all if it's too much of a pain, and they will disappear all by themselves.  In the meantime let's design for 64-bit platforms and let's not waste time on 32-bit corner cases.

gasche commented:
I understand this as "not in the mood to discuss an Int64.t on the OCaml side". I'm going to do the prudent thing and add a weird hack to avoid 0 on overflow, to keep the change minimal.

gasche committed:
domain: avoid reusing 0 on domain unique_id wraparound

gasche commented:
I just pushed a new commit with the avoid-0 behavior on wraparound. I would welcome an extra pair of eyes before merging.

If someone is interested in changing the `Domain.id` type to be `Int64.t` instead of `int`, please support my suggestion *now*! It's probably not going to be possible to change it after a 5.00 release.

xavierleroy commented on runtime/domain.c:
```diff
@@ -389,6 +389,24 @@ int caml_reallocate_minor_heap(asize_t wsize)
+static uintnat next_domain_unique_id = 0;
+
+/* Precondition: you must own [all_domains_lock]. */
+static uintnat fresh_domain_unique_id(void) {
+    uintnat next = next_domain_unique_id++;
+
+    /* On 32-bit systems, there is a risk of wraparound of the unique
+       id counter. We have decided to let that happen and live with
+       it, but we still ensure that id 0 is not reused, to avoid
+       having new domains believe that they are the main domain. */
+    if (next_domain_unique_id == 0)
+      next_domain_unique_id++;
+
+    return next;
+}
```
You're doing a post-increment, so the first time `fresh_domain_unique_id` is called  it will return... 0 !

Either you initialize `next_domain_unique_id` with 1 or you write this:
```suggestion
static uintnat fresh_domain_unique_id(void) {
    next_domain_unique_id++;

    /* On 32-bit systems, there is a risk of wraparound of the unique
       id counter. We have decided to let that happen and live with
       it, but we still ensure that id 0 is not reused, to avoid
       having new domains believe that they are the main domain. */
    if (next_domain_unique_id == 0) next_domain_unique_id = 1;

    return next_domain_unique_id;
}
```

gasche replied:
This is by design: the first domain is initialized by a call to `create_domain` in `caml_init_domains`, and I want this very first call to return 0.

https://github.com/ocaml/ocaml/blob/6a7a7b36fa2b6f4c4f3d210c8ecaff6faa094adf/runtime/domain.c#L618-L619

I guess that this should be made more explicit, possibly by doing an assert on `domain_self`'s unique id right after initialization.

xavierleroy replied:
Ouch... that's a fragile design!  You'd better document it, then.

gasche replied:
This is documented now and tested by a `CAMLassert` and in the testsuite.

gasche replied:
(I'm secretely hoping that the "Changes requested" status will be changed back to "merge merge merge!".)

xavierleroy replied:
Thanks for the added comments.  Will remove the mark of infamy ASAP.

xavierleroy reviewed (changes requested):
Something looks broken in the computation of the domain IDs.

Otherwise, +1 for keeping `int` as the type of domain IDs.

gasche committed:
update tests/parallel/domain_id.ml

gasche commented:
@ctk21 

> I also checked if tests are dependent on the implementation of domain id. `tests/parallel/domain_id.ml` relies on monotonic increasing ids and so is robust to the change; however it has a comment about `Max_domains` that should be updated to match. (There are also usages of `Domain.self` in tests with ignored return!)

Thanks for pointing this out. I looked at the test, and I believe that it is broken: it is supposed to check that domain slots get reused as expected, but there is no robust way to do this without having access to the slot index, and the approximation used in the test tries to reason about unique_ids in a clever way... that doesn't work anymore, yet the test still passes.

For now I have removed this test, and added instead a check that `Domain.is_main_domain` works when called from the main domain.

ctk21 commented:
> Thanks for pointing this out. I looked at the test, and I believe that it is broken: it is supposed to check that domain slots get reused as expected, but there is no robust way to do this without having access to the slot index, and the approximation used in the test tries to reason about unique_ids in a clever way... that doesn't work anymore, yet the test still passes.

Agreed, that delta based test is not correct with the new implementation. Your removal is sensible.

xavierleroy reviewed (approved):
Looks good to me.

gasche commented:
Thanks everyone, and in particular @kayceesrk, @xavierleroy and @ctk21 for your excellent reviews.

(I had a discussion with @dra27 about Multicore-related changes entries. Our mutual understanding is that small increments over the Multicore runtime (not visible as distinct behavior changes to 4.x users) may not deserve an entry in general -- we may still want to record credits when the people doing the work are new or uncommon contributors. Larger features and work of course often deserve their own entry. I classified this specific change in the "small increment" category.)
